File: lib/admin/features/users/services/admin_user_service.dart
import 'package:cloud_firestore/cloud_firestore.dart' as fs;

import '../../../../core/models/enums.dart';
import '../../../../core/models/user_model.dart';
import '../../../../core/services/firebase_service.dart';
import '../../../core/admin_audit_service.dart';

class UserPage {
  final List<AppUser> users;
  final fs.DocumentSnapshot<Map<String, dynamic>>? lastDoc;
  final bool hasMore;
  const UserPage({required this.users, required this.lastDoc, required this.hasMore});
}

/// Admin's view of the `users` collection: browse/filter/paginate, search,
/// and moderate. All reads here are only possible because `firestore.rules`
/// grants `isAdmin()` an unrestricted read on `users/{userId}` — nothing
/// special happens in this service to earn that, it's just what the
/// signed-in account's ID token already carries.
class AdminUserService {
  AdminUserService._();
  static final AdminUserService instance = AdminUserService._();

  fs.CollectionReference<Map<String, dynamic>> get _col =>
      FirebaseService.instance.firestore.collection('users');
  fs.CollectionReference<Map<String, dynamic>> get _publicProfilesCol =>
      FirebaseService.instance.firestore.collection('providerPublicProfiles');

  /// Browse, filtered by role and/or account status, newest first.
  Future<UserPage> fetchPage({
    required UserRole? role,
    AccountStatus? status,
    fs.DocumentSnapshot<Map<String, dynamic>>? startAfter,
    int limit = 25,
  }) async {
    fs.Query<Map<String, dynamic>> query = _col;
    if (role != null) {
      query = query.where('role', isEqualTo: role.name);
      if (status != null) {
        query = query.where('accountStatus', isEqualTo: status.name);
      }
      query = query.orderBy('createdAt', descending: true);
    } else {
      // No role filter: `createdAt` alone needs no composite index. A status
      // filter with no role filter would (`accountStatus` + `createdAt`
      // without `role` isn't one of the indexes declared), so status is only
      // honored when a role is also chosen — the UI disables it otherwise.
      query = query.orderBy('createdAt', descending: true);
    }
    if (startAfter != null) {
      query = query.startAfterDocument(startAfter);
    }
    final snap = await query.limit(limit + 1).get();
    final docs = snap.docs;
    final hasMore = docs.length > limit;
    final pageDocs = hasMore ? docs.sublist(0, limit) : docs;
    return UserPage(
      users: pageDocs.map((d) => AppUser.fromMap(d.data())).toList(),
      lastDoc: pageDocs.isEmpty ? null : pageDocs.last,
      hasMore: hasMore,
    );
  }

  /// Exact-email or name-prefix search. Firestore has no full-text search
  /// without a third-party index (Algolia/Typesense) — this is the honest
  /// boundary of what a single-field range/equality query can do. A prefix
  /// search on `name` is done with the standard Firestore range trick: end
  /// the range at the search term with a high sentinel codepoint appended,
  /// so it sorts after every string that starts with the term.
  Future<List<AppUser>> search(String term) async {
    final trimmed = term.trim();
    if (trimmed.isEmpty) return [];
    if (trimmed.contains('@')) {
      final snap = await _col
          .where('email', isEqualTo: trimmed.toLowerCase())
          .limit(20)
          .get();
      return snap.docs.map((d) => AppUser.fromMap(d.data())).toList();
    }
    final rangeEnd = '$trimmed';
    final snap = await _col
        .orderBy('name')
        .startAt([trimmed])
        .endAt([rangeEnd])
        .limit(20)
        .get();
    return snap.docs.map((d) => AppUser.fromMap(d.data())).toList();
  }

  Future<AppUser?> fetchOne(String userId) async {
    final doc = await _col.doc(userId).get();
    if (!doc.exists || doc.data() == null) return null;
    return AppUser.fromMap(doc.data()!);
  }

  /// Editable-by-admin fields only — a deliberately smaller set than what
  /// `firestore.rules` would technically allow `isAdmin()` to write, so a
  /// slip in this screen can't silently rewrite `role` or `createdAt`.
  Future<void> updateProfile(AppUser user, {String? name, String? phone}) async {
    await _col.doc(user.id).update({
      'name': ?name,
      'phone': ?phone,
      'updatedAt': DateTime.now().toIso8601String(),
    });
    await AdminAuditService.instance.log(
      action: 'user.update_profile',
      targetType: 'user',
      targetId: user.id,
      details: {'name': ?name, 'phone': ?phone},
    );
  }

  Future<void> setAccountStatus(
    AppUser user,
    AccountStatus status, {
    required String reason,
  }) async {
    await _col.doc(user.id).update({
      'accountStatus': status.name,
      'updatedAt': DateTime.now().toIso8601String(),
    });
    await AdminAuditService.instance.log(
      action: 'user.set_account_status',
      targetType: 'user',
      targetId: user.id,
      details: {'status': status.name, 'reason': reason, 'email': user.email},
    );
  }

  /// Grants or revokes the Verified badge. The provider's own device
  /// republishes `providerPublicProfiles/{uid}` to match on its next launch
  /// — see `ProviderDirectoryService.publishSelf` in the mobile app and
  /// `docs/ADMIN_DATA_CONTRACT.md`.
  Future<void> setVerified(AppUser user, bool verified) async {
    await _col.doc(user.id).update({
      'isVerified': verified,
      'updatedAt': DateTime.now().toIso8601String(),
    });
    await AdminAuditService.instance.log(
      action: verified ? 'user.verify' : 'user.unverify',
      targetType: 'user',
      targetId: user.id,
      details: {'email': user.email},
    );
  }

  /// Deletes the Firestore profile (and, for a provider, their public
  /// profile mirror). **Does not delete the Firebase Auth account** —
  /// nothing client-side can; that needs the Admin SDK. Run
  /// `node scripts/admin/delete_user.js <email>` afterwards to remove the
  /// login itself — see `scripts/admin/README.md`.
  Future<void> deleteProfile(AppUser user) async {
    await _col.doc(user.id).delete();
    if (user.role == UserRole.provider) {
      await _publicProfilesCol.doc(user.id).delete().catchError((_) {});
    }
    await AdminAuditService.instance.log(
      action: 'user.delete_profile',
      targetType: 'user',
      targetId: user.id,
      details: {'email': user.email, 'role': user.role.name},
    );
  }
}

File: admin-web/src/services/adminUserService.ts
import {
  collection, doc, getDoc, getDocs, deleteDoc, updateDoc,
  query, where, orderBy, startAt, endAt, startAfter, limit as fsLimit,
  type QueryDocumentSnapshot, type DocumentData,
} from 'firebase/firestore'
import { db } from '@/lib/firebase'
import { logAuditEvent } from './adminAuditService'
import { mapAppUser, type AppUser } from '@/types/models'
import type { AccountStatus, UserRole } from '@/types/enums'

const usersCol = () => collection(db, 'users')
const publicProfilesCol = () => collection(db, 'providerPublicProfiles')

export interface UserPage {
  users: AppUser[]
  lastDoc: QueryDocumentSnapshot<DocumentData> | null
  hasMore: boolean
}

/** Browse, filtered by role and/or account status, newest first. Status is only honored when a role is also chosen (mirrors the index constraint in AdminUserService.fetchPage). */
export async function fetchUserPage(opts: {
  role?: UserRole
  status?: AccountStatus
  startAfterDoc?: QueryDocumentSnapshot<DocumentData>
  pageLimit?: number
}): Promise<UserPage> {
  const limitValue = opts.pageLimit ?? 25
  const constraints = []
  if (opts.role) {
    constraints.push(where('role', '==', opts.role))
    if (opts.status) constraints.push(where('accountStatus', '==', opts.status))
  }
  constraints.push(orderBy('createdAt', 'desc'))
  if (opts.startAfterDoc) constraints.push(startAfter(opts.startAfterDoc))
  constraints.push(fsLimit(limitValue + 1))

  const snap = await getDocs(query(usersCol(), ...constraints))
  const docs = snap.docs
  const hasMore = docs.length > limitValue
  const pageDocs = hasMore ? docs.slice(0, limitValue) : docs
  return {
    users: pageDocs.map((d) => mapAppUser(d.id, d.data())),
    lastDoc: pageDocs.length ? pageDocs[pageDocs.length - 1] : null,
    hasMore,
  }
}

/** Exact-email or name-prefix search — the honest boundary of a single-field Firestore query without a third-party index. */
export async function searchUsers(term: string): Promise<AppUser[]> {
  const trimmed = term.trim()
  if (!trimmed) return []
  if (trimmed.includes('@')) {
    const snap = await getDocs(query(usersCol(), where('email', '==', trimmed.toLowerCase()), fsLimit(20)))
    return snap.docs.map((d) => mapAppUser(d.id, d.data()))
  }
  // Standard Firestore prefix-range trick: end the range at the term with a
  // high sentinel codepoint appended so it sorts after every string that
  // starts with the term.
  const snap = await getDocs(
    query(usersCol(), orderBy('name'), startAt(trimmed), endAt(trimmed + ''), fsLimit(20)),
  )
  return snap.docs.map((d) => mapAppUser(d.id, d.data()))
}

export async function fetchUser(userId: string): Promise<AppUser | null> {
  const snap = await getDoc(doc(usersCol(), userId))
  return snap.exists() ? mapAppUser(snap.id, snap.data()) : null
}

/** Editable-by-admin fields only, deliberately smaller than firestore.rules technically allows. */
export async function updateUserProfile(user: AppUser, fields: { name?: string; phone?: string }): Promise<void> {
  await updateDoc(doc(usersCol(), user.id), {
    ...(fields.name !== undefined ? { name: fields.name } : {}),
    ...(fields.phone !== undefined ? { phone: fields.phone } : {}),
    updatedAt: new Date().toISOString(),
  })
  await logAuditEvent({
    action: 'user.update_profile',
    targetType: 'user',
    targetId: user.id,
    details: fields,
  })
}

export async function setUserAccountStatus(user: AppUser, status: AccountStatus, reason: string): Promise<void> {
  await updateDoc(doc(usersCol(), user.id), {
    accountStatus: status,
    updatedAt: new Date().toISOString(),
  })
  await logAuditEvent({
    action: 'user.set_account_status',
    targetType: 'user',
    targetId: user.id,
    details: { status, reason, email: user.email },
  })
}

/** Grants/revokes the Verified badge; the provider's own device republishes providerPublicProfiles on next launch. */
export async function setUserVerified(user: AppUser, verified: boolean): Promise<void> {
  await updateDoc(doc(usersCol(), user.id), {
    isVerified: verified,
    updatedAt: new Date().toISOString(),
  })
  await logAuditEvent({
    action: verified ? 'user.verify' : 'user.unverify',
    targetType: 'user',
    targetId: user.id,
    details: { email: user.email },
  })
}

/** Deletes the Firestore profile only — cannot delete the Firebase Auth account (needs the Admin SDK). Run scripts/admin/delete_user.js afterwards. */
export async function deleteUserProfile(user: AppUser): Promise<void> {
  await deleteDoc(doc(usersCol(), user.id))
  if (user.role === 'provider') {
    try {
      await deleteDoc(doc(publicProfilesCol(), user.id))
    } catch {
      // best-effort, mirrors the Dart .catchError((_) {})
    }
  }
  await logAuditEvent({
    action: 'user.delete_profile',
    targetType: 'user',
    targetId: user.id,
    details: { email: user.email, role: user.role },
  })
}
